<a href="https://colab.research.google.com/github/OfsTr/PythonWorks/blob/master/Bayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Bayes teoremi
# Bir araştırmaya göre her 43 çocuktan 1 tanesi, yetişkinlikte ortaya çıkan belli bir
# hastalığa yakalanmakta ve tam güvenilir olmamasına rağmen yapılan test sonuçlarına göre,
# hastalıklı bir çocuğun testi %80 pozitif, sağlıklı bir çocuğun testi ise %10 pozitif sonuç vermektedir.
# Bu bilgilere göre test sonucu pozitif olan bir çocuğun gerçekten hasta olma olasılığı nedir?

# bayes formula P(A|B)=P(B|A)*P(A)/P(B)

# kullanıcı girişi ile int döner
def getTotalUserValue():
    userinput = input("Toplam kişi sayısını giriniz : ")

    while 1 == 1:
        try:
            val = int(userinput)
            break
        except ValueError:
            userinput = input("Lütfen sadece sayı girişi yapınız : ")

    return val

# kullanıcı girişi ile int döner
def getIllUserValue():
    userinput = input("Hasta kişi sayısını giriniz : ")

    while 1 == 1:
        try:
            val = int(userinput)
            break
        except ValueError:
            userinput = input("Lütfen sadece sayı girişi yapınız : ")

    return val

# kullanıcı girişi ile rakam döner
def getIllUserPositivePercentValue():
    userinput = input("Hasta birisinin testinini pozitif çıkma olasılı %100 : ")

    while 1 == 1:
        try:
            val = int(userinput)
            break
        except ValueError:
            userinput = input("Lütfen sadece 1-100 arasında bir sayı girişi yapınız : ")

    return val

# kullanıcı girişi ile rakam döner
def getHealthyUserPositivePercentValue():
    userinput = input("Sağlıklı birisinin testinini pozitif çıkma olasılı %100 : ")

    while 1 == 1:
        try:
            val = int(userinput)
            break
        except ValueError:
            userinput = input("Lütfen sadece 1-100 arasında bir sayı girişi yapınız : ")

    return val

#toplam kişi sayısını alıyoruz
totalUserCount = getTotalUserValue()

# hasta kişi sayısını alıyoruz
illUserCount = getIllUserValue()

#hasta birisinin test sonucunun pozitif çıkması ihtimalini alıyoruz
illUserPositiveTestPercent = getIllUserPositivePercentValue() / 100

#sağlıklı birisinin testinin pozitif çıkma ihtimalini alıyoruz
healtyUserPositiveTestPercent = getHealthyUserPositivePercentValue() / 100

#kişinin hasta ve sağlıklı olma ihtimallerini hesaplıyoruz
probabltyIllUser = illUserCount / totalUserCount
probabltyHealtyUser = 1 - probabltyIllUser

#yapılan testin herhangi birisinde pozitif sonuç verme ihtimalini hesaplıyoruz
probabltyPositiveTestPercent = probabltyIllUser * illUserPositiveTestPercent + probabltyHealtyUser * healtyUserPositiveTestPercent

#testi pozitif çıkan birisinin gerçekten hasta olma ihtimalini hesaplıyoruz
result = illUserPositiveTestPercent * probabltyIllUser / probabltyPositiveTestPercent

print("Testin pozitif çıkma olasılığı : " + str(probabltyPositiveTestPercent))

print("Kişinin hasta olma olaslığı : " + str(probabltyIllUser))

print("Test sonucu pozitif olan birisinin gerçekten hasta olma ihtimali :" + str(result))



Toplam kişi sayısını giriniz : 43
Hasta kişi sayısını giriniz : 1
Hasta birisinin testinini pozitif çıkma olasılı %100 : 80
Sağlıklı birisinin testinini pozitif çıkma olasılı %100 : 10
Testin pozitif çıkma olasılığı : 0.11627906976744186
Kişinin hasta olma olaslığı : 0.023255813953488372
Test sonucu pozitif olan birisinin gerçekten hasta olma ihtimali :0.16


** BAYES TEOREMİ ÖRNEK**
 
 Problem : Bir araştırmaya göre her 43 çocuktan 1 tanesi, yetişkinlikte ortaya çıkan belli bir hastalığa yakalanmakta ve tam güvenilir olmamasına rağmen yapılan test sonuçlarına göre, hastalıklı bir çocuğun testi %80 pozitif, sağlıklı bir çocuğun testi ise %10 pozitif sonuç vermektedir. Bu bilgilere göre test sonucu pozitif olan bir çocuğun gerçekten hasta olma olasılığı nedir?
 
 Yukarıda ki problemi parametreler kullanıcıdan alınacak şekilde sonucu hesaplayacak programı geliştiriyoruz.
 Öncelikle kullanıcıdan paremetleri okumak için dört adet method yazıyoruz ; 

**  1. getTotalUserValue():**
    Toplam kişi sayısı kullanıcıdan alınıyor.
    
**  2. getIllUserValue():**
    Toplam kişiler içersinde kaçının hasta olduğu bilgisi kullanıcıdan alınıyor.
   
**  3. getIllUserPositivePercentValue():**
    Hasta birisine yapılan bir testin pozitif çıkma ihtimali kullanıcıdan alınıyor.
  
**  4. getHealthyUserPositivePercentValue():**
     Sağlıklı birisine yapılan testin pozitif çıkma ihtimali bilgisi kullanıcıdan alınıyor.

Sonrasında yapılan testin herhangi birisinde pozitif sonuç verme ihtimalini hesaplıyoruz.

Sonrasında ise bayes formülü ile değerleri yerine koyup testi pozitif çıkan birisinin hasta olma ihtimalini hesaplıyoruz.

P(A) : Çocuğun hasta olması olasılığı = 1/43
P(B) : Testin pozitif çıkması olasılığı = 1/43 * 0.80 + 42/43 * 0.10 = 5/43
P(A|B) : Pozitif çıkan testin hastalık çıkma olasılığı ( sorulan bu )
P(B|A) : Hastalıklı çocuğun testinin pozitif çıkma olasılığı = 0.80

P(A|B)=P(B|A)*P(A)/P(B) => (0.80 * 1/43) / (5/43) = 0.16 = %16 bulunur.

 